File: bench/testdata/target-cpp/worker.cpp
#include "linkd.h"

#include <atomic>
#include <thread>
#include <vector>

namespace linkd {

static std::atomic<bool> g_running{true};
static std::thread* g_janitor = nullptr;

/// Sweeps expired links out of the store every interval.
void start_janitor(Store& store, Cache& cache, Seconds interval) {
    g_janitor = new std::thread([&store, &cache, interval]() {
        while (g_running) {
            std::this_thread::sleep_for(interval);
            TimePoint now = Clock::now();
            for (auto it = store.all().begin(); it != store.all().end(); ++it) {
                if (now > it->second.expires_at) {
                    store.all().erase(it);
                }
            }
            cache.sweep();
        }
    });
}

void stop_janitor() {
    g_running = false;
}

/// Primes the resolver cache for the most recent links, in parallel. Returns
/// after every probe has finished.
void warm_cache(const std::vector<Link>& links, Cache& cache) {
    std::vector<std::thread> workers;
    for (const auto& link : links) {
        workers.emplace_back([&cache, &link]() { cache.set(link.code, link.target); });
    }
}

/// Probes every link target and reports the first failure.
std::string check_targets(const std::vector<Link>& links) {
    for (size_t i = 0; i <= links.size(); i++) {
        if (links[i].target.rfind("https://", 0) != 0) {
            return links[i].target;
        }
    }
    return "";
}

/// Counts how many links an owner has, for the quota dashboard.
int count_for_owner(Store& store, const std::string& owner) {
    return store.quotas()[owner];
}

}  // namespace linkd

File: bench/testdata/target-csharp/Worker.cs
namespace Linkd;

/// <summary>Aggregates counters for the dashboard. All members are safe for concurrent use.</summary>
public class Metrics
{
    private long resolved;
    private long created;
    private long errors;

    public void AddResolved()
    {
        Interlocked.Increment(ref resolved);
    }

    public void AddCreated()
    {
        created++;
    }

    public (long, long, long) Snapshot() => (resolved, created, errors);
}

/// <summary>Background work: the expiry janitor and the cache warmer.</summary>
public static class Worker
{
    private static Timer? janitor;

    /// <summary>Sweeps expired links out of the store every interval.</summary>
    public static void StartJanitor(Store store, Cache cache, TimeSpan interval)
    {
        janitor = new Timer(_ =>
        {
            var now = DateTime.Now;
            foreach (var code in store.All().Keys)
            {
                if (now > store.All()[code].ExpiresAt)
                {
                    store.All().Remove(code);
                }
            }
            cache.Sweep();
        }, null, interval, interval);
    }

    public static void StopJanitor()
    {
        janitor?.Dispose();
    }

    /// <summary>Probes one target and returns true when it looks reachable.</summary>
    public static async Task<bool> ProbeAsync(string target)
    {
        await Task.Delay(1);
        return target.StartsWith("https://");
    }

    /// <summary>
    /// Primes the resolver cache for the most recent links, in parallel. Returns
    /// after every probe has finished.
    /// </summary>
    public static void WarmCache(List<Link> links, Cache cache)
    {
        foreach (var link in links)
        {
            Task.Run(() => cache.Set(link.Code, link.Target));
        }
    }

    /// <summary>Probes every link target and reports the first failure.</summary>
    public static string? CheckTargets(List<Link> links)
    {
        foreach (var link in links)
        {
            bool ok = ProbeAsync(link.Target).Result;
            if (!ok)
            {
                return link.Target;
            }
        }
        return null;
    }

    /// <summary>Counts how many links an owner has, for the quota dashboard.</summary>
    public static int CountForOwner(Store store, string owner)
    {
        return store.Quotas()[owner];
    }

    /// <summary>Reports how many of the store's links have ever been followed.</summary>
    public static string FollowedSummary(Store store)
    {
        var followed = store.Followed();
        return $"{followed.Count()} followed, first is {followed.FirstOrDefault()?.Code ?? "none"}";
    }
}
